### PR TITLE
Rewrite this module

### PR DESCRIPTION
**Pull request by stevemao**
Setter. While it is possible to impliment setter using recursion, it would not work with getter (`t.assert(dotProp(f1, 'foo', 2) === 2);` would fail).
It avoids recursion. I personally find recursive functions are hard to follow.
`dotProp({foo: {bar: a}}, foo.fake.fake2)` should throw a TypeError instead of returning `undefined`.
Use lodash to avoid reinventing the wheel.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,15 +1,28 @@
 'use strict';
+var each = require('lodash.foreach');
```
Comment (sindresorhus) [inline index.js]: Array.forEach()

File: index.js [hunk 2]
```diff
@@ -1,15 +1,28 @@
...
+var isUndefined = require('lodash.isundefined');
```
Comment (sindresorhus) [inline index.js]: `=== undefined`

File: index.js [hunk 3]
```diff
@@ -1,15 +1,28 @@
...
 
-function isObject(x) {
-	return typeof x === 'object' && x !== null;
-}
+module.exports = function dotProp(obj, path, value) {
```
Comment (sindresorhus) [inline index.js]: `module.exports = function (obj, path, value) {`

File: index.js [hunk 4]
```diff
@@ -1,15 +1,28 @@
...
 
-module.exports = function getProp(obj, path) {
-	if (!isObject(obj) || typeof path !== 'string') {
-		return obj;
-	}
+  var argsLen = arguments.length;
 
-	path = path.split('.');
+  var pathArr = path.split('.');
+  each(pathArr, function(path, index) {
```
Comment (sindresorhus) [inline index.js]: `each(pathArr, function (path, index) {`

**Comment by sindresorhus** [comment]
Wonder if it makes more sense to use separate get/set methods for clarity. `get()` `set()`.

**Comment by stevemao** [comment]
I like the way to differentiate them based on `arguments.length` and this is pretty common.

**Comment by stevemao** [comment]
Done @sindresorhus

**Comment by stevemao** [comment]
Anything I can do to get this merged quickly? @sindresorhus

File: readme.md [hunk 5]
```diff
@@ -15,11 +15,17 @@ $ npm install --save dot-prop
 ```js
 var dotProp = require('dot-prop');
 
+// getter
 dotProp({foo: {bar: 'unicorn'}}, 'foo.bar');
 //=> 'unicorn'
 
-dotProp({foo: {bar: 'a'}}, 'foo.notDefined.deep');
```
Comment (floatdrop) [inline readme.md]: Is this a regression? Can you add this test:

```js
t.assert(dotProp(f1, 'foo.baz.bar') === undefined);
```

Comment (stevemao) [inline readme.md]: Why? There is a test
```
t.assert(dotProp({foo: {bar: {baz: null}}}, 'foo.bar.baz') === null);
```

Comment (floatdrop) [inline readme.md]: There is a difference: `'foo.baz.bar'` vs `'foo.bar.baz'`

**Comment by arthurvr** [comment]
@stevemao Why did you change the indentation?

File: readme.md (hunk 5 above)
Comment (stevemao) [inline readme.md]: This should throw an error because `foo.bar` is `undefined`. If `dotProp({foo: {bar: {baz: null}}}, 'foo.baz.bar') === undefined` it implies `foo.baz` is not `undefined`

Comment (floatdrop) [inline readme.md]: But this is essential part of this plugin - https://github.com/sindresorhus/dot-prop#usage and we are using it heavily.

**Comment by stevemao** [comment]
Oh sorry, will change it back

File: readme.md (hunk 5 above)
Comment (arthurvr) [inline readme.md]: yup ^

Comment (stevemao) [inline readme.md]: But how can I handle the error then? Sometimes you can't ignore it.

Comment (floatdrop) [inline readme.md]: I think, if you want an error, you could just type `f1.foo.baz.bar` - bam! error.

Comment (stevemao) [inline readme.md]: I can't... I need to use dot prop to get a value on an object. If there is an error I might need to know.
@floatdrop According to your theory, you could just type `f1.foo` - bam! You get the value.

Comment (floatdrop) [inline readme.md]: @stevemao theories... theories... anyway this is clearly a regression - a lot of people may use this module because of this feature. Throwing an exception should be discussed in separate PR.

Comment (stevemao) [inline readme.md]: feature of what? ignoring an error? There is a regression only if you don't handle the error properly.
```js
function (err, data) {
  if (err) throw err;
  console.log(data);
}
```
(taken from https://nodejs.org/api/fs.html#fs_fs_readfile_filename_options_callback)
This has a regression too but it's a design pattern of node. 

@floatdrop I'm not trying to oppose you but I don't see any use case of ignoring the error and you haven't given me one. I was expecting a good reason (with references maybe).

https://www.joyent.com/developers/node/design/errors

Comment (floatdrop) [inline readme.md]: @stevemao I give'd you one https://github.com/lapwinglabs/rube/blob/1c8af1a0f48c9f27b048a2beadcf2df739605f42/lib/object.js#L85-L87

We have code like this in production:

```js
var translation = prop(i18n, 'some.keyset.key') || defaultValue;
```

We are perfectly aware, that some keys may be absent and want to get default value instead of `try...catching` errors. I think this is pretty valid use case.

Comment (stevemao) [inline readme.md]: K :+1:

Comment (sindresorhus) [inline readme.md]: The whole point of this module is not throwing on undefined sub-keys.

File: index.js [hunk 6]
```diff
@@ -1,15 +1,26 @@
...
 
-module.exports = function getProp(obj, path) {
-	if (!isObject(obj) || typeof path !== 'string') {
-		return obj;
-	}
+  var argsLen = arguments.length;
 
-	path = path.split('.');
+  var pathArr = path.split('.');
+  pathArr.forEach(function(path, index) {
```
Comment (sindresorhus) [inline index.js]: `pathArr.forEach(function (path, index) {`

File: index.js [hunk 7]
```diff
@@ -1,15 +1,26 @@
 'use strict';
 
-function isObject(x) {
-	return typeof x === 'object' && x !== null;
-}
+module.exports = function(obj, path, value) {
```
Comment (sindresorhus) [inline index.js]: `module.exports = function (obj, path, value) {`

**Comment by sindresorhus** [comment]
> I like the way to differentiate them based on arguments.length and this is pretty common.

Common doesn't mean it's a good pattern. I don't like functions that completely change behaviour based on function arity. Much better to be explicit. Can you change it into `.get()` and `.set()` methods?

**Comment by stevemao** [comment]
@sindresorhus  Happy to change it. But I don't think the behaviours are completely different since you could get the value if you are setting it at the same time. https://github.com/stevemao/dot-prop/commit/9f3f16c91e10f817413b17a231b6ee6407205b33#diff-1dd241c4cd3fd1dd89c570cee98b79ddR27 You could think this is "get the value and change to a new one"

If you just pass two arguments you could think this is "get the value and don't change it"

It doesn't matter how many arguments present it's always both get and set. Kinda like `Array.prototype.splice()`. Again this is just my understanding. Will change this soon.

**Comment by sindresorhus** [comment]
> Happy to change it.

Yes, please :)

We could argue back and forth about the merits. In the end it's mostly preference. I prefer to keep my methods explicit and non-magic. This also makes it easier to scan the code as it's abundantly clear what it does. `dotProp(f1, 'foo', 2)` without any context is not.

> But I don't think the behaviours are completely different since you could get the value if you are setting it at the same time.

I see no usefulness for that as it will just give you the exact same value as you passed in.

**Comment by floatdrop** [comment]
@stevemao any updates?

**Comment by stevemao** [comment]
Sorry, I'm on this now.

**Comment by stevemao** [comment]
Done

**Comment by floatdrop** [comment]
Sorry for be a douche again, but what should happen in this situation?

```js
dotProp.set(f1, 'foo.bar', 2);
dotProp.set(f1, 'foo.bar.baz', 3);
t.assert(f1.foo.bar.baz === 3); // f1.foo.bar is 2, so f1.foo.bar.baz is undefined
```

**Comment by stevemao** [comment]
:+1: fixed

File: test.js [hunk 8]
```diff
@@ -2,16 +2,54 @@
...
+
+	dotProp.set(f1, 'foo', 2);
+	t.assert(f1.foo === 2);
+
+	f1 = {foo: {bar: 1}};
+	dotProp.set(f1, 'foo.bar', 2);
+	t.assert(f1.foo.bar === 2);
+
+	dotProp.set(f1, 'foo.bar.baz', 3);
+	t.assert(f1.foo.bar.baz === 3); // f1.foo.bar is 2, so f1.foo.bar.baz is undefined
```
Comment (floatdrop) [inline test.js]: Remove this comment, plz.

Comment (stevemao) [inline test.js]: yep

**Comment by floatdrop** [comment]
lgtm :star2:

File: index.js [hunk 9]
```diff
@@ -4,12 +4,40 @@ function isObject(x) {
...
+
+	return obj;
+};
+
+module.exports.set = function (obj, path, value) {
+	if (!isObject(obj) || typeof path !== 'string') {
+		return;
+	}
+
+	var argsLen = arguments.length;
```
Comment (floatdrop) [inline index.js]: unused variable

**Comment by sindresorhus** [comment]
Very nice work @stevemao :)

**Comment by stevemao** [comment]
Thanks :D
